File: src/main/java/com/cxxcxx/zinecraft/api/localization/TranslationCatalog.java
package com.cxxcxx.zinecraft.api.localization;

import java.util.LinkedHashMap;
import java.util.regex.Pattern;

public final class TranslationCatalog {
  private static final Pattern KEY = Pattern.compile("[A-Za-z0-9_.:/-]+");
  private final LinkedHashMap<String, LocalizedText> entries = new LinkedHashMap<>();

  public static void addWithDefaults(TranslationCatalog self, String key, String zhCn, String enUs, int mask, Object marker) {
    self.add(key, zhCn, (mask & 4) != 0 ? zhCn : enUs);
  }

  public static void replaceWithDefaults(TranslationCatalog self, String key, String zhCn, String enUs, int mask, Object marker) {
    self.replace(key, zhCn, (mask & 4) != 0 ? zhCn : enUs);
  }

  public LinkedHashMap<String, LocalizedText> getEntries() {
    return entries;
  }

  public void add(String key, String zhCn, String enUs) {
    var text = validatedText(key, zhCn, enUs);
    var previous = entries.putIfAbsent(key, text);
    if (previous != null && !previous.equals(text)) throw new IllegalArgumentException("翻译键重复且内容冲突：" + key);
  }

  public void replace(String key, String zhCn, String enUs) {
    var text = validatedText(key, zhCn, enUs);
    if (!entries.containsKey(key)) throw new IllegalArgumentException("不能替换尚未注册的翻译键：" + key);
    entries.put(key, text);
  }

  private LocalizedText validatedText(String key, String zhCn, String enUs) {
    if (!KEY.matcher(key).matches()) throw new IllegalArgumentException("翻译键格式无效：" + key);
    if (zhCn.isBlank()) throw new IllegalArgumentException("中文翻译不能为空：" + key);
    if (enUs.isBlank()) throw new IllegalArgumentException("英文翻译不能为空：" + key);
    return new LocalizedText(zhCn, enUs);
  }
}

File: src/main/kotlin/com/cxxcxx/zinecraft/api/localization/TranslationCatalog.kt
package com.cxxcxx.zinecraft.api.localization

/** Collects localized text that is emitted by the language data providers. */
class TranslationCatalog {
  internal val entries = linkedMapOf<String, LocalizedText>()

  fun add(key: String, zhCn: String, enUs: String = zhCn) {
    val text = validatedText(key, zhCn, enUs)
    val previous = entries.putIfAbsent(key, text)
    require(previous == null || previous == text) { "翻译键重复且内容冲突：$key；需要覆盖时请显式调用 replace" }
  }

  /** 显式替换已注册翻译，避免拼写错误把普通新增静默变成覆盖。 */
  fun replace(key: String, zhCn: String, enUs: String = zhCn) {
    val text = validatedText(key, zhCn, enUs)
    require(entries.containsKey(key)) { "不能替换尚未注册的翻译键：$key" }
    entries[key] = text
  }

  private fun validatedText(key: String, zhCn: String, enUs: String): LocalizedText {
    require(key.matches(Regex("[A-Za-z0-9_.:/-]+"))) { "翻译键格式无效：$key" }
    require(zhCn.isNotBlank()) { "中文翻译不能为空：$key" }
    require(enUs.isNotBlank()) { "英文翻译不能为空：$key" }
    return LocalizedText(zhCn, enUs)
  }
}

data class LocalizedText(val zhCn: String, val enUs: String)

internal fun String.toDisplayName(): String =
  split('_', '.').filter(String::isNotEmpty).joinToString(" ") { word ->
    word.replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
  }
